crates/cargo-test-support/src/compare.rs

```diff
@@ -706,6 +669,136 @@ impl fmt::Debug for WildStr<'_> {
     }
 }
 
+pub struct InMemoryDir {
+    files: Vec<(PathBuf, Data)>,
+}
+
+impl InMemoryDir {
+    pub fn paths(&self) -> impl Iterator<Item = &Path> {
+        self.files.iter().map(|(p, _)| p.as_path())
+    }
+
+    #[track_caller]
+    pub fn assert_contains(&self, expected: &Self) {
+        use std::fmt::Write as _;
+        let assert = assert_e2e();
+        let mut errs = String::new();
+        for (path, expected_data) in &expected.files {
+            let actual_data = self
+                .files
+                .iter()
+                .find_map(|(p, d)| (path == p).then(|| d.clone()))
+                .unwrap_or_else(|| Data::new());
+            if let Err(err) =
+                assert.try_eq(Some(&path.display()), actual_data, expected_data.clone())
+            {
+                let _ = write!(&mut errs, "{err}");
+            }
+        }
+        if !errs.is_empty() {
+            panic!("{errs}")
+        }
+    }
+}
+
+impl<P, D> FromIterator<(P, D)> for InMemoryDir
+where
+    P: Into<std::path::PathBuf>,
+    D: IntoData,
+{
+    fn from_iter<I: IntoIterator<Item = (P, D)>>(files: I) -> Self {
+        let files = files
+            .into_iter()
+            .map(|(p, d)| (p.into(), d.into_data()))
+            .collect();
+        Self { files }
+    }
+}
+
+impl<const N: usize, P, D> From<[(P, D); N]> for InMemoryDir
+where
+    P: Into<PathBuf>,
+    D: IntoData,
+{
+    fn from(files: [(P, D); N]) -> Self {
+        let files = files
+            .into_iter()
+            .map(|(p, d)| (p.into(), d.into_data()))
+            .collect();
+        Self { files }
+    }
+}
+
+impl<P, D> From<std::collections::HashMap<P, D>> for InMemoryDir
+where
+    P: Into<PathBuf>,
+    D: IntoData,
+{
+    fn from(files: std::collections::HashMap<P, D>) -> Self {
+        let files = files
+            .into_iter()
+            .map(|(p, d)| (p.into(), d.into_data()))
+            .collect();
+        Self { files }
+    }
+}
+
+impl<P, D> From<std::collections::BTreeMap<P, D>> for InMemoryDir
+where
+    P: Into<PathBuf>,
+    D: IntoData,
+{
+    fn from(files: std::collections::BTreeMap<P, D>) -> Self {
+        let files = files
+            .into_iter()
+            .map(|(p, d)| (p.into(), d.into_data()))
+            .collect();
+        Self { files }
+    }
+}
+
+impl From<()> for InMemoryDir {
+    fn from(_files: ()) -> Self {
+        let files = Vec::new();
+        Self { files }
+    }
+}
+
+macro_rules! impl_from_tuple_for_dirsnapshot {
```

Comment:
I know it is working, and I understand it is for supporting up to 9 tuples (right?). However the macro is a bit hard to read for me. Is this a de-facto decl macro pattern? Is it possible to document what it is doing?

Reply:
Thanks!

`r=me` when the version bump is fixed